lib/src/main/java/graphql/nadel/engine/transform/hydration/batch/NadelNewBatchHydrator.kt

```diff
@@ -497,6 +528,26 @@ internal class NadelNewBatchHydrator(
         }
     }
 
+    private fun getFilteredInstructions(
+        instructions: List<NadelBatchHydrationFieldInstruction>,
+        // fieldSource: ValueSource.FieldResultValue,
+        sourceObject: JsonNode,
+        state: State,
+    ): List<NadelBatchHydrationFieldInstruction> {
+
+        if (instructions.any { it.condition == null }) {
+            return instructions
+        }
+
+        return instructions.filter {
+            // Note: due to the validation, all instructions in here have a condition, so can call explicitly
+            val resultQueryPath = state.aliasHelper.getQueryPath(it.condition!!.fieldPath)
+            val node = JsonNodeExtractor.getNodesAt(sourceObject, resultQueryPath)
+                .emptyOrSingle()
+            it.condition.evaluate(node?.value)
+        }.firstOrNull()?.let { listOf(it) } ?: emptyList()
```

Comment:
One thing, this should be like the other class, so you should be doing `.firstOrNull {condition.evaluate }`

But also, you should look into Kotlin sequences which is like Java streams. You should avoid calling transform functions directly on lists because it then generates another list, which means more allocations and slower execution.

i.e. you are currently invoking

```
public inline fun <T> Iterable<T>.filter(predicate: (T) -> Boolean): List<T>
```

The return type is a new list.

Reply:
changed, now using `.firstOrNull {condition.evaluate }`